
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="csvtk - CSV/TSV toolkit">
      
      
        <meta name="author" content="shenwei356">
      
      
      
      
        <link rel="next" href="download/">
      
      
      <link rel="icon" href="files/favicon.ico">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.4.8">
    
    
      
        <title>csvtk - CSV/TSV Toolkit</title>
      
    
    
      <link rel="stylesheet" href="assets/stylesheets/main.4b4a2bd9.min.css">
      
        
        <link rel="stylesheet" href="assets/stylesheets/palette.356b1318.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="light-blue" data-md-color-accent="light-blue">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#csvtk-a-cross-platform-efficient-and-practical-csvtsv-toolkit" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="." title="csvtk - CSV/TSV Toolkit" class="md-header__button md-logo" aria-label="csvtk - CSV/TSV Toolkit" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            csvtk - CSV/TSV Toolkit
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Home
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/shenwei356/csvtk" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="." title="csvtk - CSV/TSV Toolkit" class="md-nav__button md-logo" aria-label="csvtk - CSV/TSV Toolkit" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    csvtk - CSV/TSV Toolkit
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/shenwei356/csvtk" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Home
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="." class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#introduction" class="md-nav__link">
    Introduction
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#table-of-contents" class="md-nav__link">
    Table of Contents
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#features" class="md-nav__link">
    Features
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#subcommands" class="md-nav__link">
    Subcommands
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#installation" class="md-nav__link">
    Installation
  </a>
  
    <nav class="md-nav" aria-label="Installation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#method-1-download-binaries-latest-stabledev-version" class="md-nav__link">
    Method 1: Download binaries (latest stable/dev version)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-2-install-via-conda-latest-stable-version" class="md-nav__link">
    Method 2: Install via conda (latest stable version)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-3-install-via-homebrew" class="md-nav__link">
    Method 3: Install via homebrew
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-4-for-go-developer-latest-stabledev-version" class="md-nav__link">
    Method 4: For Go developer (latest stable/dev version)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-5-for-archlinux-aur-users-may-be-not-the-latest" class="md-nav__link">
    Method 5: For ArchLinux AUR users (may be not the latest)
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#command-line-completion" class="md-nav__link">
    Command-line completion
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#compared-to-csvkit" class="md-nav__link">
    Compared to csvkit
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#examples" class="md-nav__link">
    Examples
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#acknowledgements" class="md-nav__link">
    Acknowledgements
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#contact" class="md-nav__link">
    Contact
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#license" class="md-nav__link">
    License
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#starchart" class="md-nav__link">
    Starchart
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="download/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Download
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="usage/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Usage
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="faq/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    FAQs
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="tutorial/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Tutorial
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="chinese/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    中文介绍
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
    <li class="md-nav__item">
      <a href="https://github.com/shenwei356" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    More tools
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#introduction" class="md-nav__link">
    Introduction
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#table-of-contents" class="md-nav__link">
    Table of Contents
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#features" class="md-nav__link">
    Features
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#subcommands" class="md-nav__link">
    Subcommands
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#installation" class="md-nav__link">
    Installation
  </a>
  
    <nav class="md-nav" aria-label="Installation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#method-1-download-binaries-latest-stabledev-version" class="md-nav__link">
    Method 1: Download binaries (latest stable/dev version)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-2-install-via-conda-latest-stable-version" class="md-nav__link">
    Method 2: Install via conda (latest stable version)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-3-install-via-homebrew" class="md-nav__link">
    Method 3: Install via homebrew
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-4-for-go-developer-latest-stabledev-version" class="md-nav__link">
    Method 4: For Go developer (latest stable/dev version)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-5-for-archlinux-aur-users-may-be-not-the-latest" class="md-nav__link">
    Method 5: For ArchLinux AUR users (may be not the latest)
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#command-line-completion" class="md-nav__link">
    Command-line completion
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#compared-to-csvkit" class="md-nav__link">
    Compared to csvkit
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#examples" class="md-nav__link">
    Examples
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#acknowledgements" class="md-nav__link">
    Acknowledgements
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#contact" class="md-nav__link">
    Contact
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#license" class="md-nav__link">
    License
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#starchart" class="md-nav__link">
    Starchart
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="csvtk-a-cross-platform-efficient-and-practical-csvtsv-toolkit">csvtk - a cross-platform, efficient and practical CSV/TSV toolkit</h1>
<ul>
<li><strong>Documents:</strong> <a href="http://bioinf.shenwei.me/csvtk/">http://bioinf.shenwei.me/csvtk</a>
( <a href="http://bioinf.shenwei.me/csvtk/usage/"><strong>Usage</strong></a>, <a href="http://bioinf.shenwei.me/csvtk/tutorial/"><strong>Tutorial</strong></a> and <a href="http://bioinf.shenwei.me/csvtk/faq/"><strong>FAQs</strong></a>).
<a href="http://bioinf.shenwei.me/csvtk/chinese">中文介绍</a></li>
<li><strong>Source code:</strong>  <a href="https://github.com/shenwei356/csvtk">https://github.com/shenwei356/csvtk</a> <a href="https://github.com/shenwei356/csvtk"><img alt="GitHub stars" src="https://img.shields.io/github/stars/shenwei356/csvtk.svg?style=social&amp;label=Star&amp;?maxAge=2592000" /></a>
<a href="https://github.com/shenwei356/csvtk/blob/master/LICENSE"><img alt="license" src="https://img.shields.io/github/license/shenwei356/csvtk.svg?maxAge=2592000" /></a></li>
<li><strong>Latest version:</strong> <a href="https://github.com/shenwei356/csvtk/releases"><img alt="Latest Stable Version" src="https://img.shields.io/github/release/shenwei356/csvtk.svg?style=flat" /></a>
<a href="http://bioinf.shenwei.me/csvtk/download/"><img alt="Github Releases" src="https://img.shields.io/github/downloads/shenwei356/csvtk/latest/total.svg?maxAge=3600" /></a>
<a href="http://bioinf.shenwei.me/csvtk/download/"><img alt="Cross-platform" src="https://img.shields.io/badge/platform-any-ec2eb4.svg?style=flat" /></a>
<a href="https://anaconda.org/bioconda/csvtk"><img alt="Anaconda Cloud" src="https://anaconda.org/bioconda/csvtk/badges/version.svg" /></a></li>
</ul>
<h2 id="introduction">Introduction</h2>
<p>Similar to FASTA/Q format in field of Bioinformatics,
CSV/TSV formats are basic and ubiquitous file formats in both Bioinformatics and data science.</p>
<p>People usually use spreadsheet software like MS Excel to process table data.
However this is all by clicking and typing, which is <strong>not
automated and is time-consuming to repeat</strong>, especially when you want to
apply similar operations with different datasets or purposes.</p>
<p><strong><em>You can also accomplish some CSV/TSV manipulations using shell commands,
but more code is needed to handle the header line. Shell commands do not
support selecting columns with column names either.</em></strong></p>
<p><code>csvtk</code> is <strong>convenient for rapid data investigation
and also easy to integrate into analysis pipelines</strong>.
It could save you lots of time in (not) writing Python/R scripts.</p>
<h2 id="table-of-contents">Table of Contents</h2>
<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

<ul>
<li><a href="#features">Features</a></li>
<li><a href="#subcommands">Subcommands</a></li>
<li><a href="#installation">Installation</a></li>
<li><a href="#command-line-completion">Command-line completion</a></li>
<li><a href="#compared-to-csvkit">Compared to <code>csvkit</code></a></li>
<li><a href="#examples">Examples</a></li>
<li><a href="#acknowledgements">Acknowledgements</a></li>
<li><a href="#contact">Contact</a></li>
<li><a href="#license">License</a></li>
<li><a href="#starchart">Starchart</a></li>
</ul>
<!-- END doctoc generated TOC please keep comment here to allow auto update -->

<h2 id="features">Features</h2>
<ul>
<li><strong>Cross-platform</strong> (Linux/Windows/Mac OS X/OpenBSD/FreeBSD)</li>
<li><strong>Light weight and out-of-the-box, no dependencies, no compilation, no configuration</strong></li>
<li><strong>Fast</strong>,  <strong>multiple-CPUs supported</strong> (some commands)</li>
<li><strong>Practical functions provided by N subcommands</strong></li>
<li><strong>Support STDIN and gziped input/output file, easy being used in pipe</strong></li>
<li>Most of the subcommands support <strong><em>unselecting fields</em></strong> and <strong><em>fuzzy fields</em></strong>,
  e.g. <code>-f "-id,-name"</code> for all fields except "id" and "name",
  <code>-F -f "a.*"</code> for all fields with prefix "a.".</li>
<li><strong>Support some common plots</strong> (see <a href="http://bioinf.shenwei.me/csvtk/usage/#plot">usage</a>)</li>
<li><del>Seamlessly support for data with meta line (e.g., <code>sep=,</code>) of separator declaration used by MS Excel</del></li>
</ul>
<h2 id="subcommands">Subcommands</h2>
<p>53 subcommands in total.</p>
<p><strong>Information</strong></p>
<ul>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#headers"><code>headers</code></a>: prints headers</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#dim/nrow/ncol"><code>dim</code></a>: dimensions of CSV file</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#dim/nrow/ncol"><code>nrow</code></a>: print number of records</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#dim/nrow/ncol"><code>ncol</code></a>: print number of columns</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#summary"><code>summary</code></a>: summary statistics of selected numeric or text fields (groupby group fields)</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#watch"><code>watch</code></a>: online monitoring and histogram of selected field</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#corr"><code>corr</code></a>: calculate Pearson correlation between numeric columns</li>
</ul>
<p><strong>Format conversion</strong></p>
<ul>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#pretty"><code>pretty</code></a>: converts CSV to a readable aligned table</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#csv2tab"><code>csv2tab</code></a>: converts CSV to tabular format</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#tab2csv"><code>tab2csv</code></a>: converts tabular format to CSV</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#space2tab"><code>space2tab</code></a>: converts space delimited format to TSV</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#csv2md"><code>csv2md</code></a>: converts CSV to markdown format</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#csv2rst"><code>csv2rst</code></a>: converts CSV to reStructuredText format</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#csv2json"><code>csv2json</code></a>: converts CSV to JSON format</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#csv2xlsx"><code>csv2xlsx</code></a>: converts CSV/TSV files to XLSX file</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#xlsx2csv"><code>xlsx2csv</code></a>: converts XLSX to CSV format</li>
</ul>
<p><strong>Set operations</strong></p>
<ul>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#head"><code>head</code></a>: prints first N records</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#concat"><code>concat</code></a>: concatenates CSV/TSV files by rows</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#sample"><code>sample</code></a>: sampling by proportion</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#cut"><code>cut</code></a>: select and arrange fields</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#grep"><code>grep</code></a>: greps data by selected fields with patterns/regular expressions</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#uniq"><code>uniq</code></a>: unique data without sorting</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#freq"><code>freq</code></a>: frequencies of selected fields</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#inter"><code>inter</code></a>: intersection of multiple files</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#filter"><code>filter</code></a>: filters rows by values of selected fields with arithmetic expression</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#filter2"><code>filter2</code></a>: filters rows by awk-like arithmetic/string expressions</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#join"><code>join</code></a>: join files by selected fields (inner, left and outer join)</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#split"><code>split</code></a> splits CSV/TSV into multiple files according to column values</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#splitxlsx"><code>splitxlsx</code></a>: splits XLSX sheet into multiple sheets according to column values</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#comb"><code>comb</code></a>: compute combinations of items at every row</li>
</ul>
<p><strong>Edit</strong></p>
<ul>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#fix"><code>fix</code></a>: fix CSV/TSV with different numbers of columns in rows</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#fix-quotes"><code>fix-quotes</code></a>: fix malformed CSV/TSV caused by double-quotes</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#del-quotes"><code>del-quotes</code></a>: remove extra double-quotes added by <code>fix-quotes</code></li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#add-header"><code>add-header</code></a>: add column names</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#del-header"><code>del-header</code></a>: delete column names</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#rename"><code>rename</code></a>: renames column names with new names</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#rename2"><code>rename2</code></a>: renames column names by regular expression</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#replace"><code>replace</code></a>: replaces data of selected fields by regular expression</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#round"><code>round</code></a>: round float to n decimal places</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#mutate"><code>mutate</code></a>: creates new columns from selected fields by regular expression</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#mutate2"><code>mutate2</code></a>: creates a new column from selected fields by awk-like arithmetic/string expressions</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#fmtdate"><code>fmtdate</code></a>: format date of selected fields</li>
</ul>
<p><strong>Transform</strong></p>
<ul>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#transpose"><code>transpose</code></a>: transposes CSV data</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#sep"><code>sep</code></a>: separate column into multiple columns</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#gather"><code>gather</code></a>: gather columns into key-value pairs, like <code>tidyr::gather/pivot_longer</code></li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#spread"><code>spread</code></a>: spread a key-value pair across multiple columns, like <code>tidyr::spread/pivot_wider</code></li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#unfold"><code>unfold</code></a>: unfold multiple values in cells of a field</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#fold"><code>fold</code></a>: fold multiple values of a field into cells of groups</li>
</ul>
<p><strong>Ordering</strong></p>
<ul>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#sort"><code>sort</code></a>: sorts by selected fields</li>
</ul>
<p><strong>Ploting</strong></p>
<ul>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#plot"><code>plot</code></a> see <a href="http://bioinf.shenwei.me/csvtk/usage/#plot">usage</a><ul>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#hist"><code>plot hist</code></a> histogram</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#box"><code>plot box</code></a> boxplot</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#line"><code>plot line</code></a> line plot and scatter plot</li>
</ul>
</li>
</ul>
<p><strong>Misc</strong></p>
<ul>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#cat"><code>cat</code></a> stream file and report progress</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#version"><code>version</code></a>   print version information and check for update</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#genautocomplete"><code>genautocomplete</code></a> generate shell autocompletion script (bash|zsh|fish|powershell)</li>
</ul>
<h2 id="installation">Installation</h2>
<p><a href="https://github.com/shenwei356/csvtk/releases">Download Page</a></p>
<p><code>csvtk</code> is implemented in <a href="https://golang.org/">Go</a> programming language,
 executable binary files <strong>for most popular operating systems</strong> are freely available
  in <a href="https://github.com/shenwei356/csvtk/releases">release</a> page.</p>
<h4 id="method-1-download-binaries-latest-stabledev-version">Method 1: Download binaries (latest stable/dev version)</h4>
<p>Just <a href="https://github.com/shenwei356/csvtk/releases">download</a> compressed
executable file of your operating system,
and decompress it with <code>tar -zxvf *.tar.gz</code> command or other tools.
And then:</p>
<ol>
<li>
<p><strong>For Linux-like systems</strong></p>
<ol>
<li>
<p>If you have root privilege simply copy it to <code>/usr/local/bin</code>:</p>
<pre><code>sudo cp csvtk /usr/local/bin/
</code></pre>
</li>
<li>
<p>Or copy to anywhere in the environment variable <code>PATH</code>:</p>
<pre><code>mkdir -p $HOME/bin/; cp csvtk $HOME/bin/
</code></pre>
</li>
</ol>
</li>
<li>
<p><strong>For windows</strong>, just copy <code>csvtk.exe</code> to <code>C:\WINDOWS\system32</code>.</p>
</li>
</ol>
<h4 id="method-2-install-via-conda-latest-stable-version">Method 2: Install via conda (latest stable version)  <a href="https://anaconda.org/bioconda/csvtk"><img alt="Anaconda Cloud" src="https://anaconda.org/bioconda/csvtk/badges/version.svg" /></a> <a href="https://anaconda.org/bioconda/csvtk"><img alt="downloads" src="https://anaconda.org/bioconda/csvtk/badges/downloads.svg" /></a></h4>
<pre><code>conda install -c bioconda csvtk
</code></pre>
<h4 id="method-3-install-via-homebrew">Method 3: Install via homebrew</h4>
<pre><code>brew install csvtk
</code></pre>
<h4 id="method-4-for-go-developer-latest-stabledev-version">Method 4: For Go developer (latest stable/dev version)</h4>
<pre><code>go get -u github.com/shenwei356/csvtk/csvtk
</code></pre>
<h4 id="method-5-for-archlinux-aur-users-may-be-not-the-latest">Method 5: For ArchLinux AUR users (may be not the latest)</h4>
<pre><code>yaourt -S csvtk
</code></pre>
<h2 id="command-line-completion">Command-line completion</h2>
<p>Bash:</p>
<pre><code># generate completion shell
csvtk genautocomplete --shell bash

# configure if never did.
# install bash-completion if the "complete" command is not found.
echo "for bcfile in ~/.bash_completion.d/* ; do source \$bcfile; done" &gt;&gt; ~/.bash_completion
echo "source ~/.bash_completion" &gt;&gt; ~/.bashrc
</code></pre>
<p>Zsh:</p>
<pre><code># generate completion shell
csvtk genautocomplete --shell zsh --file ~/.zfunc/_csvtk

# configure if never did
echo 'fpath=( ~/.zfunc "${fpath[@]}" )' &gt;&gt; ~/.zshrc
echo "autoload -U compinit; compinit" &gt;&gt; ~/.zshrc
</code></pre>
<p>fish:</p>
<pre><code>csvtk genautocomplete --shell fish --file ~/.config/fish/completions/csvtk.fish
</code></pre>
<h2 id="compared-to-csvkit">Compared to <code>csvkit</code></h2>
<p><a href="http://csvkit.readthedocs.org/">csvkit</a>, attention: this table wasn't updated for many years.</p>
<table>
<thead>
<tr>
<th style="text-align: left;">Features</th>
<th style="text-align: center;">csvtk</th>
<th style="text-align: center;">csvkit</th>
<th style="text-align: left;">Note</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">Read    Gzip</td>
<td style="text-align: center;">Yes</td>
<td style="text-align: center;">Yes</td>
<td style="text-align: left;">read gzip files</td>
</tr>
<tr>
<td style="text-align: left;">Fields ranges</td>
<td style="text-align: center;">Yes</td>
<td style="text-align: center;">Yes</td>
<td style="text-align: left;">e.g. <code>-f 1-4,6</code></td>
</tr>
<tr>
<td style="text-align: left;"><strong>Unselect fileds</strong></td>
<td style="text-align: center;">Yes</td>
<td style="text-align: center;">--</td>
<td style="text-align: left;">e.g. <code>-1</code> for excluding first column</td>
</tr>
<tr>
<td style="text-align: left;"><strong>Fuzzy fields</strong></td>
<td style="text-align: center;">Yes</td>
<td style="text-align: center;">--</td>
<td style="text-align: left;">e.g. <code>ab*</code> for columns with name prefix "ab"</td>
</tr>
<tr>
<td style="text-align: left;">Reorder fields</td>
<td style="text-align: center;">Yes</td>
<td style="text-align: center;">Yes</td>
<td style="text-align: left;">it means <code>-f 1,2</code> is different from <code>-f 2,1</code></td>
</tr>
<tr>
<td style="text-align: left;"><strong>Rename columns</strong></td>
<td style="text-align: center;">Yes</td>
<td style="text-align: center;">--</td>
<td style="text-align: left;">rename with new name(s) or from existed names</td>
</tr>
<tr>
<td style="text-align: left;">Sort by multiple keys</td>
<td style="text-align: center;">Yes</td>
<td style="text-align: center;">Yes</td>
<td style="text-align: left;">bash sort like operations</td>
</tr>
<tr>
<td style="text-align: left;"><strong>Sort by number</strong></td>
<td style="text-align: center;">Yes</td>
<td style="text-align: center;">--</td>
<td style="text-align: left;">e.g. <code>-k 1:n</code></td>
</tr>
<tr>
<td style="text-align: left;"><strong>Multiple sort</strong></td>
<td style="text-align: center;">Yes</td>
<td style="text-align: center;">--</td>
<td style="text-align: left;">e.g. <code>-k 2:r -k 1:nr</code></td>
</tr>
<tr>
<td style="text-align: left;">Pretty output</td>
<td style="text-align: center;">Yes</td>
<td style="text-align: center;">Yes</td>
<td style="text-align: left;">convert CSV to readable aligned table</td>
</tr>
<tr>
<td style="text-align: left;"><strong>Unique data</strong></td>
<td style="text-align: center;">Yes</td>
<td style="text-align: center;">--</td>
<td style="text-align: left;">unique data of selected fields</td>
</tr>
<tr>
<td style="text-align: left;"><strong>frequency</strong></td>
<td style="text-align: center;">Yes</td>
<td style="text-align: center;">--</td>
<td style="text-align: left;">frequencies of selected fields</td>
</tr>
<tr>
<td style="text-align: left;"><strong>Sampling</strong></td>
<td style="text-align: center;">Yes</td>
<td style="text-align: center;">--</td>
<td style="text-align: left;">sampling by proportion</td>
</tr>
<tr>
<td style="text-align: left;"><strong>Mutate fields</strong></td>
<td style="text-align: center;">Yes</td>
<td style="text-align: center;">--</td>
<td style="text-align: left;">create new columns from selected fields</td>
</tr>
<tr>
<td style="text-align: left;"><strong>Replace</strong></td>
<td style="text-align: center;">Yes</td>
<td style="text-align: center;">--</td>
<td style="text-align: left;">replace data of selected fields</td>
</tr>
</tbody>
</table>
<p>Similar tools:</p>
<ul>
<li><a href="http://csvkit.readthedocs.org/">csvkit</a> - A suite of utilities for converting to and working with CSV, the king of tabular file formats. http://csvkit.rtfd.org/</li>
<li><a href="https://github.com/BurntSushi/xsv">xsv</a> - A fast CSV toolkit written in Rust.</li>
<li><a href="https://github.com/johnkerl/miller">miller</a> - Miller is like sed, awk, cut, join, and sort for
name-indexed data such as CSV and tabular JSON http://johnkerl.org/miller</li>
<li><a href="https://github.com/eBay/tsv-utils">tsv-utils</a> - Command line utilities for tab-separated value files written in the D programming language.</li>
</ul>
<h2 id="examples">Examples</h2>
<p>More <a href="http://shenwei356.github.io/csvtk/usage/">examples</a> and <a href="http://shenwei356.github.io/csvtk/tutorial/">tutorial</a>.</p>
<p><strong>Attention</strong></p>
<ol>
<li>By default, csvtk assumes input files have header row, if not, switch flag <code>-H</code> on.</li>
<li>By default, csvtk handles CSV files, use flag <code>-t</code> for tab-delimited files.</li>
<li>Column names should be unique.</li>
<li>By default, lines starting with <code>#</code> will be ignored, if the header row
    starts with <code>#</code>, please assign flag <code>-C</code> another rare symbol, e.g. <code>$</code>.</li>
<li>Do not mix use field (column) numbers and names to specify columns to operate.</li>
<li>The CSV parser requires all the lines have same numbers of fields/columns.
    Even lines with spaces will cause error.
    Use <code>-I/--ignore-illegal-row</code> to skip these lines if neccessary.
    You can also use "csvtk fix" to fix files with different numbers of columns in rows.</li>
<li>
<p>If double-quotes exist in fields not enclosed with double-quotes, e.g.,</p>
<pre><code>x,a "b" c,1
</code></pre>
<p>It would report error:</p>
<pre><code>bare `"` in non-quoted-field.
</code></pre>
<p>Please switch on the flag <code>-l</code> or use <code>csvtk fix-quotes</code> to fix it.</p>
</li>
<li>
<p>If somes fields have only a double-quote eighter in the beginning or in the end, e.g.,</p>
<pre><code>x,d "e","a" b c,1
</code></pre>
<p>It would report error:</p>
<pre><code>extraneous or missing " in quoted-field
</code></pre>
<p>Please use <code>csvtk fix-quotes</code> to fix it, and use <code>csvtk del-quotes</code> to reset to the
original format as needed.</p>
</li>
</ol>
<p>Examples</p>
<ol>
<li>
<p>Pretty result</p>
<pre><code>$ csvtk pretty names.csv
id   first_name   last_name   username
--   ----------   ---------   --------
11   Rob          Pike        rob
2    Ken          Thompson    ken
4    Robert       Griesemer   gri
1    Robert       Thompson    abc
NA   Robert       Abel        123

$ csvtk pretty names.csv -S 3line
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
 id   first_name   last_name   username
----------------------------------------
 11   Rob          Pike        rob
 2    Ken          Thompson    ken
 4    Robert       Griesemer   gri
 1    Robert       Thompson    abc
 NA   Robert       Abel        123
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

$ csvtk pretty names.csv -S bold -w 5 -m 1-
┏━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━┓
┃  id   ┃ first_name ┃ last_name ┃ username ┃
┣━━━━━━━╋━━━━━━━━━━━━╋━━━━━━━━━━━╋━━━━━━━━━━┫
┃  11   ┃    Rob     ┃   Pike    ┃   rob    ┃
┣━━━━━━━╋━━━━━━━━━━━━╋━━━━━━━━━━━╋━━━━━━━━━━┫
┃   2   ┃    Ken     ┃ Thompson  ┃   ken    ┃
┣━━━━━━━╋━━━━━━━━━━━━╋━━━━━━━━━━━╋━━━━━━━━━━┫
┃   4   ┃   Robert   ┃ Griesemer ┃   gri    ┃
┣━━━━━━━╋━━━━━━━━━━━━╋━━━━━━━━━━━╋━━━━━━━━━━┫
┃   1   ┃   Robert   ┃ Thompson  ┃   abc    ┃
┣━━━━━━━╋━━━━━━━━━━━━╋━━━━━━━━━━━╋━━━━━━━━━━┫
┃  NA   ┃   Robert   ┃   Abel    ┃   123    ┃
┗━━━━━━━┻━━━━━━━━━━━━┻━━━━━━━━━━━┻━━━━━━━━━━┛
</code></pre>
</li>
<li>
<p>Summary of selected numeric fields, supporting "group-by"</p>
<pre><code>$ cat testdata/digitals2.csv \
    | csvtk summary -i -f f4:sum,f5:sum -g f1,f2 \
    | csvtk pretty
f1    f2     f4:sum   f5:sum
bar   xyz    7.00     106.00
bar   xyz2   4.00     4.00
foo   bar    6.00     3.00
foo   bar2   4.50     5.00
</code></pre>
</li>
<li>
<p>Select fields/columns (<code>cut</code>)</p>
<ul>
<li>By index: <code>csvtk cut -f 1,2</code></li>
<li>By names: <code>csvtk cut -f first_name,username</code></li>
<li><strong>Unselect</strong>: <code>csvtk cut -f -1,-2</code> or <code>csvtk cut -f -first_name</code></li>
<li><strong>Fuzzy fields</strong>: <code>csvtk cut -F -f "*_name,username"</code></li>
<li>Field ranges: <code>csvtk cut -f 2-4</code> for column 2,3,4 or <code>csvtk cut -f -3--1</code> for discarding column 1,2,3</li>
<li>All fields: <code>csvtk cut -f 1-</code> or  <code>csvtk cut -F -f "*"</code></li>
</ul>
</li>
<li>
<p>Search by selected fields (<code>grep</code>) (matched parts will be highlighted as red)</p>
<ul>
<li>By exactly matching: <code>csvtk grep -f first_name -p Robert -p Rob</code></li>
<li>By regular expression: <code>csvtk grep -f first_name -r -p Rob</code></li>
<li>By pattern list: <code>csvtk grep -f first_name -P name_list.txt</code></li>
<li>Remore rows containing missing data (NA): <code>csvtk grep -F -f "*" -r -p "^$" -v</code></li>
</ul>
</li>
<li>
<p><strong>Rename column names</strong> (<code>rename</code> and <code>rename2</code>)</p>
<ul>
<li>Setting new names: <code>csvtk rename -f A,B -n a,b</code> or <code>csvtk rename -f 1-3 -n a,b,c</code></li>
<li>Replacing with original names by regular express: <code>csvtk rename2 -f 1- -p "(.*)" -r 'prefix_$1'</code> for adding prefix to all column names.</li>
</ul>
</li>
<li>
<p><strong>Edit data with regular expression</strong> (<code>replace</code>)</p>
<ul>
<li>Remove Chinese charactors:  <code>csvtk replace -F -f "*_name" -p "\p{Han}+" -r ""</code></li>
</ul>
</li>
<li>
<p><strong>Create new column from selected fields by regular expression</strong> (<code>mutate</code>)</p>
<ul>
<li>In default, copy a column: <code>csvtk mutate -f id</code></li>
<li>Extract prefix of data as group name (get "A" from "A.1" as group name):
  <code>csvtk mutate -f sample -n group -p "^(.+?)\." --after sample</code></li>
</ul>
</li>
<li>
<p>Sort by multiple keys (<code>sort</code>)</p>
<ul>
<li>By single column : <code>csvtk sort -k 1</code> or <code>csvtk sort -k last_name</code></li>
<li>By multiple columns: <code>csvtk sort -k 1,2</code> or <code>csvtk sort -k 1 -k 2</code> or <code>csvtk sort -k last_name,age</code></li>
<li>Sort by number: <code>csvtk sort -k 1:n</code> or  <code>csvtk sort -k 1:nr</code> for reverse number</li>
<li>Complex sort: <code>csvtk sort -k region -k age:n -k id:nr</code></li>
<li>In natural order: <code>csvtk sort -k chr:N</code></li>
</ul>
</li>
<li>
<p><strong>Join multiple files by keys</strong> (<code>join</code>)</p>
<ul>
<li>All files have same key column: <code>csvtk join -f id file1.csv file2.csv</code></li>
<li>Files have different key columns: <code>csvtk join -f "username;username;name" names.csv phone.csv adress.csv -k</code></li>
</ul>
</li>
<li>
<p>Filter by numbers (<code>filter</code>)</p>
<ul>
<li>Single field: <code>csvtk filter -f "id&gt;0"</code></li>
<li><strong>Multiple fields</strong>: <code>csvtk filter -f "1-3&gt;0"</code></li>
<li>Using <code>--any</code> to print record if any of the field satisfy the condition: <code>csvtk filter -f "1-3&gt;0" --any</code></li>
<li><strong>fuzzy fields</strong>: <code>csvtk filter -F -f "A*!=0"</code></li>
</ul>
</li>
<li>
<p><strong>Filter rows by awk-like arithmetic/string expressions</strong> (<code>filter2</code>)</p>
<ul>
<li>Using field index: <code>csvtk filter2 -f '$3&gt;0'</code></li>
<li>Using column names: <code>csvtk filter2 -f '$id &gt; 0'</code></li>
<li>Both arithmetic and string expressions: <code>csvtk filter2 -f '$id &gt; 3 || $username=="ken"'</code></li>
<li>More complicated: <code>csvtk filter2 -H -t -f '$1 &gt; 2 &amp;&amp; $2 % 2 == 0'</code></li>
</ul>
</li>
<li>
<p>Ploting</p>
<ul>
<li>plot histogram with data of the second column:<pre><code>csvtk -t plot hist testdata/grouped_data.tsv.gz -f 2 | display
</code></pre>
</li>
</ul>
<p><img alt="histogram.png" src="testdata/figures/histogram.png" /></p>
<ul>
<li>plot boxplot with data of the "GC Content" (third) column,
group information is the "Group" column.<pre><code>csvtk -t plot box testdata/grouped_data.tsv.gz -g "Group" \
    -f "GC Content" --width 3 --title "Box plot" | display
</code></pre>
</li>
</ul>
<p><img alt="boxplot.png" src="testdata/figures/boxplot.png" /></p>
<ul>
<li>plot horiz boxplot with data of the "Length" (second) column,
group information is the "Group" column.<pre><code>csvtk -t plot box testdata/grouped_data.tsv.gz -g "Group" -f "Length"  \
    --height 3 --width 5 --horiz --title "Horiz box plot" | display
</code></pre>
</li>
</ul>
<p><img alt="boxplot2.png" src="testdata/figures/boxplot2.png" /></p>
<ul>
<li>plot line plot with X-Y data<pre><code>csvtk -t plot line testdata/xy.tsv -x X -y Y -g Group | display
</code></pre>
</li>
</ul>
<p><img alt="lineplot.png" src="testdata/figures/lineplot.png" /></p>
<ul>
<li>plot scatter plot with X-Y data<pre><code>csvtk -t plot line testdata/xy.tsv -x X -y Y -g Group --scatter | display
</code></pre>
</li>
</ul>
<p><img alt="scatter.png" src="testdata/figures/scatter.png" /></p>
</li>
</ol>
<h2 id="acknowledgements">Acknowledgements</h2>
<p>We are grateful to <a href="https://github.com/dawnmy">Zhiluo Deng</a> and
<a href="https://github.com/penglbio">Li Peng</a> for suggesting features and reporting bugs.</p>
<p>Thanks <a href="https://github.com/avilella">Albert Vilella</a> for features suggestion,
which makes csvtk feature-rich。</p>
<h2 id="contact">Contact</h2>
<p><a href="https://github.com/shenwei356/csvtk/issues">Create an issue</a> to report bugs,
propose new functions or ask for help.</p>
<p>Or <a href="https://shenwei356.github.io/csvtk/usage/#disqus_thread">leave a comment</a>.</p>
<h2 id="license">License</h2>
<p><a href="https://github.com/shenwei356/csvtk/blob/master/LICENSE">MIT License</a></p>
<h2 id="starchart">Starchart</h2>
<p><img src="https://starchart.cc/shenwei356/csvtk.svg" alt="Stargazers over time" style="max-width: 100%"></p>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": ".", "features": [], "search": "assets/javascripts/workers/search.f886a092.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="assets/javascripts/bundle.81fa17fe.min.js"></script>
      
    
  </body>
</html>